sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala

```diff
@@ -683,6 +683,12 @@ class SparkSession private(
     ret
   }
 
+  def observation(expr: Column, exprs: Column*): Observation =
```

Comment:
Should this be a DataFrame API? A common pattern in pyspark is `df.select(df.my_col)`, but it's weird to see `spark.observation(df.my_col)`.

Reply:
This method is to create the `Observation` instance, whereas a method on a DataFrame would be expected to return a DataFrame, as in `Dataset[T].observe(name, expr, exprs): Dataset[T]`.

This approach here is similar to AccumulatorV2, which is instantiated and registered by calling a method on `SparkContext`, e.g. `SparkContext.LongAccumulator`.

An alternative would extend the Dataset API as you suggest:

    val observation = Observation(count(lit(1)))
    df.observe(observation).collect()
    val count = observation.get

Reply:
It's a bit awkward to build an `Observation` with columns. The `Observation` looks like can be used with different DataFrames, but it is not because the columns are from a certain DataFrame.

How about this:
```
val observation = Observation("the name")
df1.observe(observation, expressions...)...
val res1 = observation.get

df2.observe(observation, expressions...)...
val res2 = observation.get
```

Reply:
> because the columns are from a certain DataFrame

As long as the columns exist in the DataFrame, you can use it for any DataFrame. Column names should resolve when used on a specific DataFrame with compatible schema.

```
Observation(count($"id"), sum($"downloads"))
```

With your approach you could use the observation with varying columns. If you see Observation merely as a container to retrieve arbitrary results from `df.observe`, then your API makes most sense. If an Observation *is* its aggregation expressions, that is then applied on multiple compatible DataFrames, the suggested API is more conicse.

Reply:
Yea this is arguable. Personally, I'd like to see a new API similar to the existing `df.observe(...)`, but the new one here is very different:
```
val ob = spark.observe(...)
df.transform(ob.on).
```

Let's get more feedbacks. cc @srowen @HyukjinKwon @dongjoon-hyun @maropu